### PR TITLE
UAF-58 / UAF-2426 MOD-PA1031-05 eInvoicing

### PR DESCRIPTION
**Pull request by adamkost**
This pull request was left as 2 commits intentionally.
The first commit contains many methods and xml beans copied from KFS6 to a new/existing edu.arizona child counterpart file. This was done to highlight the MANY changes that were necessary.

The second commit contains the changes made to the appropriate methods and the additional code to support those changes.

That being said, if the reviewer would prefer to have this squashed into a single commit, that can be done.

### CODE REVIEW

File: kfs-core/src/main/java/edu/arizona/kfs/sys/service/impl/TaxServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    @Override
+    public List<TaxDetail> getSalesTaxDetails(Date dateOfTransaction, String postalCode, KualiDecimal amount) {
+        List<TaxDetail> salesTaxDetails = new ArrayList<TaxDetail>();
+
+        if (StringUtils.isNotBlank(postalCode)) {
+            // strip digits from the postal code before passing it to the sales tax regions
+            // if the parameters indicate to do so.
+            postalCode = truncatePostalCodeForSalesTaxRegionService(postalCode);
+            List<TaxRegion> salesTaxRegions = taxRegionService.getSalesTaxRegions(postalCode);
+            TaxDetail newTaxDetail = null;
```
Comment (sskinner) [inline kfs-core/src/main/java/edu/arizona/kfs/sys/service/impl/TaxServiceImpl.java:26]: No biggie, but can be declared in loop. (No change requested)

File: kfs-core/src/main/resources/edu/arizona/kfs/sys/businessobject/datadictionary/GenericAttributes.xml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<beans xmlns="http://www.springframework.org/schema/beans"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xmlns:p="http://www.springframework.org/schema/p"
```
Comment (sskinner) [inline kfs-core/src/main/resources/edu/arizona/kfs/sys/businessobject/datadictionary/GenericAttributes.xml:4]: p-namespace not used, could be deleted. (No change requested)

File: kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/validation/impl/ElectronicInvoiceRejectDocumentRule.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    @Override
+    protected boolean processCustomSaveDocumentBusinessRules(Document document) {
+        if (GlobalVariables.getMessageMap().hasErrors()) {
+            MessageMap errorMap = GlobalVariables.getMessageMap();
+            if (errorMap.containsMessageKey(KFSKeyConstants.ERROR_MAX_LENGTH)) {
+                return false;
+            }
+        }
+        boolean isValid = processBusinessRules(document);
```
Comment (sskinner) [inline kfs-purap/src/main/java/edu/arizona/kfs/module/purap/document/validation/impl/ElectronicInvoiceRejectDocumentRule.java:25]: We could remove the LHS, we just need the method call to set any error messages. (No changes requested)

File: kfs-purap/src/main/java/edu/arizona/kfs/module/purap/service/impl/ElectronicInvoiceHelperServiceImpl.java [hunk 4]
```diff
@@ -37,64 +52,348 @@
...
+            mailText.append(PurapConstants.ElectronicInvoice.NO_FILES_PROCESSED_EMAIL_MESSAGE);
+            mailText.append("\n\n");
 
-    protected PaymentRequestDocument createPaymentRequest(ElectronicInvoiceOrderHolder orderHolder){
+            sendSummary(mailText);
+            return eInvoiceLoad;
+        }
+
+        try {
+            /**
```
Comment (sskinner) [inline kfs-purap/src/main/java/edu/arizona/kfs/module/purap/service/impl/ElectronicInvoiceHelperServiceImpl.java:139]: Dangling javadoc (no change requested).

File: kfs-purap/src/main/java/edu/arizona/kfs/module/purap/service/impl/ElectronicInvoiceHelperServiceImpl.java [hunk 5]
```diff
@@ -37,64 +52,348 @@
...
+                     * is throwing a ValidationException if the error map is not empty before routing the doc.
+                     */
+                    GlobalVariables.getMessageMap().clearErrorMessages();
+
+                    // Do not execute rest of code below
+                    continue;
+                }
+            }
+
+            /**
```
Comment (sskinner) [inline kfs-purap/src/main/java/edu/arizona/kfs/module/purap/service/impl/ElectronicInvoiceHelperServiceImpl.java:210]: Another javadoc.

File: kfs-purap/src/main/java/edu/arizona/kfs/module/purap/service/impl/ElectronicInvoiceMatchingServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+    }
+
+    @Override
+    public void doMatchingProcess(ElectronicInvoiceOrderHolder orderHolder) {
+
+        if (LOG.isInfoEnabled()) {
+            LOG.info("Matching process started");
+        }
+
+        upperVariancePercentString = SpringContext.getBean(ParameterService.class).getParameterValueAsString(ElectronicInvoiceStep.class, PurapParameterConstants.ElectronicInvoiceParameters.SALES_TAX_UPPER_VARIANCE_PERCENT);
```
Comment (sskinner) [inline kfs-purap/src/main/java/edu/arizona/kfs/module/purap/service/impl/ElectronicInvoiceMatchingServiceImpl.java:47]: We're a service, any reason not to inject ParameterService?
(Undecided on if this is a requested change; let's see what the rest looks like.)

File: kfs-purap/src/main/java/edu/arizona/kfs/module/purap/service/impl/ElectronicInvoiceMatchingServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+        }
+
+    }
+
+    @Override
+    protected void validateQtyBasedItem(ElectronicInvoiceItemHolder itemHolder) {
+
+        PurchaseOrderItem poItem = itemHolder.getPurchaseOrderItem();
+
+        String fileName = itemHolder.getInvoiceOrderHolder().getFileName();
```
Comment (sskinner) [inline kfs-purap/src/main/java/edu/arizona/kfs/module/purap/service/impl/ElectronicInvoiceMatchingServiceImpl.java:248]: Dead variable.

File: kfs-purap/src/main/java/edu/arizona/kfs/module/purap/service/impl/ElectronicInvoiceMatchingServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,476 @@
...
+        }
+
+    }
+
+    @Override
+    protected void validateNonQtyBasedItem(ElectronicInvoiceItemHolder itemHolder) {
+
+        PurchaseOrderItem poItem = itemHolder.getPurchaseOrderItem();
+
+        String fileName = itemHolder.getInvoiceOrderHolder().getFileName();
```
Comment (sskinner) [inline kfs-purap/src/main/java/edu/arizona/kfs/module/purap/service/impl/ElectronicInvoiceMatchingServiceImpl.java:283]: Dead variable.

File: kfs-purap/src/main/java/org/kuali/kfs/module/purap/businessobject/ElectronicInvoice.java [hunk 9]
```diff
@@ -46,7 +46,8 @@
     public static String INVOICE_AMOUNT_TYPE_CODE_DISCOUNT = "DISC";
     public static String INVOICE_AMOUNT_TYPE_CODE_DEPOSIT = "DPST";
     public static String INVOICE_AMOUNT_TYPE_CODE_DUE = "DUE";
-    public static String INVOICE_AMOUNT_TYPE_CODE_EXMT = "EXMT";
+    // Note: EXMT was added in 6.0. Ted Nasser made the decision to replace EXMT with TXEX, the code used by UA in 3.
```
Comment (sskinner) [inline kfs-purap/src/main/java/org/kuali/kfs/module/purap/businessobject/ElectronicInvoice.java:49]: Thank you for this comment.

File: kfs-purap/src/main/resources/edu/arizona/kfs/module/purap/businessobject/datadictionary/ElectronicInvoiceRejectItem.xml [hunk 10]
```diff
@@ -0,0 +1,53 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<beans xmlns="http://www.springframework.org/schema/beans"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xmlns:p="http://www.springframework.org/schema/p"
+    xmlns:dd="http://rice.kuali.org/dd"
```
Comment (sskinner) [inline kfs-purap/src/main/resources/edu/arizona/kfs/module/purap/businessobject/datadictionary/ElectronicInvoiceRejectItem.xml:5]: DD-namespace not used, could be deleted.

File: kfs-purap/src/main/resources/edu/arizona/kfs/module/purap/document/datadictionary/ElectronicInvoiceRejectDocument.xml [hunk 11]
```diff
@@ -0,0 +1,233 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<beans xmlns="http://www.springframework.org/schema/beans"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xmlns:p="http://www.springframework.org/schema/p"
+    xmlns:dd="http://rice.kuali.org/dd"
```
Comment (sskinner) [inline kfs-purap/src/main/resources/edu/arizona/kfs/module/purap/document/datadictionary/ElectronicInvoiceRejectDocument.xml:5]: P/DD namespaces can be deleted.

**Review by sskinner — APPROVED** [review]
Any of the things I noted for the most part are inconsequential, and I didn't see any logical or serious issues. If we happen to revisit w/ a followup PR, might be nice to cleanup the nits. Merging, and excellent job!

**Comment by adamkost** [comment]
Many of these notes involve the copied foundation code. After testing, should there be more code changes needed, I can certainly look into making these requested changes.

**Comment by sskinner** [comment]
Hi Adam, I figured, and I consider them nits, not too consequential, which is why I merged it. I'm thinking if there has to be a tweak after the fact, like a bug being found kinda thing, then it might be nice to catch the nits at the same time. Otherwise, I don't think it's worth holding anything up over, or spending more time right now on.

**Pull request merged**